### DISCUSSION TITLE
Examples of changing a chart size on click

### DISCUSSION THREAD

**Discussion by czf** [Megabit/Blazorise#2289] [Q&A] [answered]
I'm using charts and would like to be able to change the ratio programmatically but it seems like any changes to the ratio are ignored and only the first setting is used.

**Comment by stsrki** [comment, marked as answer]
Can you show us the code how you use charts?

**Reply by czf** [reply]
Here's a simplified example, though I've tried a few different things.

I declare a pie chart on a page. 
`
<PieChart @ref="Chart" TItem="long"  OptionsObject="ChartOptionsObj"  Clicked="OnPieChartClickHandler" />
`

The default constructor for the page initializes the ChartOptionsObj.

```
        public object ChartOptionsObj { get; set; }

       public PieChartPage()
        {
            #region chartOptions
            ChartOptionsObj = new  { AspectRatio = .1 };
            #endregion chartOptions
        }
```

A method is called to reload the chart and should overwrite the AspectRatio:

```
        public async Task RefreshChartData()
        {
                  ChartOptionsObj = new
                  {
                       AspectRatio = 2
                 });
         ....
```

**Reply by stsrki** [reply]
You need to call SetOptions to notify chartjs of new settings. eg

```cs
public async Task RefreshChartData()
{
    ChartOptionsObj = new
    {
        AspectRatio = 2
    });

    await chart.SetOptions(null);
}
```

**Reply by czf** [reply]
I tried that in this branch:
https://github.com/czf/Blazorise/tree/resize
but it doesn't seem to work.  Am I still doing something wrong?

**Reply by stsrki** [reply]
I'm really not sure. I tried it myself and it is not working. Maybe the parent element is affecting it. Or maybe I'm missing some function call on chartjs instance?

**Reply by czf** [reply]
it looked like a call to `update` was missing in `setOptions`.  I tried adding that but still doesn't seem to work.  I even tried hard coding a value and nothing.  Weird that only options has this problem.  
```
    setOptions: (canvasId, options, optionsJsonString, optionsObject) => {
        if (optionsJsonString) {
            options = JSON.parse(optionsJsonString);
        }
        else if (optionsObject) {
            options = optionsObject;
        }

        const chart = window.blazoriseCharts.getChart(canvasId);

        if (chart) {
            //chart.options = options;
            chart.options.aspectRatio = 2;
            chart.update();
        }
        
    },
```

**Reply by stsrki** [reply]
Yeah, I also tried calling  `chart.update();` without any luck. I have no idea what is wrong here...

**Reply by czf** [reply]
This seems to get the size to update

```
setOptions: (canvasId, options, optionsJsonString, optionsObject) => {
        if (optionsJsonString) {
            options = JSON.parse(optionsJsonString);
        }
        else if (optionsObject) {
            options = optionsObject;
        }

        const chart = window.blazoriseCharts.getChart(canvasId);

        if (chart) {
            chart.options.aspectRatio = .1;
            chart.aspectRatio = .1;
            chart.update();
            chart.resize();
        }
        
    },
```

**Reply by stsrki** [reply]
So it works only if options are set directly, following with `update()` and `resize()` calls?

**Reply by czf** [reply]
Oh this is only as far as I got last night.  I wanted to find one case I could get working and now work backwards when I have more time to spend on it.

I think there's something funky with the way the aspectRatio is set.   This code looks like it will handle the options update for aspect ratio:
```
if (chart) {
            chart.options = options;            
            chart.aspectRatio = options.aspectRatio ?? chart.aspectRatio;
            chart.update();
            chart.resize();
        }
```

When I was running this in the debugger I noticed that the chart itself had an aspectRatio property in addition to the aspectRatio property in options.  Looks like the property inside options is ignored after initialization and the property on chart is what needs updating.  
![image](https://user-images.githubusercontent.com/2455501/117522687-90298880-af69-11eb-90f0-29c7f3e7ccb0.png)

**Reply by czf** [reply]
and now that I knew what to look for I was able to find this SO post mentioning the same thing:
https://stackoverflow.com/a/58816740

**Reply by stsrki** [reply]
Basically, it is a bug in ChartJS. Not sure do I try to add new API to be able to change this or just leave it as it is. You can always work around it with a little JSInterop and custom JavaScript code.
